From 3d1c7aee401898d76c173c3288722d2892548a73 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Fri, 7 May 2021 02:36:39 +0200
Subject: [PATCH 352/467] usb: typec: anx7688: Properly handle system suspend

Shut down the periodic timer. Remove the work from the workqueue
and wait for it to finish. Re-check the status of the anx7688 after
resume and restart the timer.

I did some suspend/resume cycle testing while plugging and removing
the PD charger at random times, and the driver seems to do the right
thing, now.

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 drivers/usb/typec/anx7688.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/drivers/usb/typec/anx7688.c b/drivers/usb/typec/anx7688.c
index db0fe6c21..bcabcc595 100644
--- a/drivers/usb/typec/anx7688.c
+++ b/drivers/usb/typec/anx7688.c
@@ -2128,6 +2128,32 @@ static int anx7688_i2c_remove(struct i2c_client *client)
         return 0;
 }
 
+static int __maybe_unused anx7688_suspend(struct device *dev)
+{
+	struct anx7688 *anx7688 = i2c_get_clientdata(to_i2c_client(dev));
+
+	del_timer_sync(&anx7688->work_timer);
+	cancel_delayed_work_sync(&anx7688->work);
+
+	return 0;
+}
+
+static int __maybe_unused anx7688_resume(struct device *dev)
+{
+	struct anx7688 *anx7688 = i2c_get_clientdata(to_i2c_client(dev));
+
+	// check status right after resume, since it could have changed during
+	// sleep
+	schedule_delayed_work(&anx7688->work, msecs_to_jiffies(50));
+	mod_timer(&anx7688->work_timer, jiffies + msecs_to_jiffies(1000));
+
+	return 0;
+}
+
+static const struct dev_pm_ops anx7688_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(anx7688_suspend, anx7688_resume)
+};
+
 static const struct i2c_device_id anx7688_ids[] = {
         { "anx7688", 0 },
         { }
@@ -2144,6 +2170,7 @@ static struct i2c_driver anx7688_driver = {
         .driver = {
                 .name = "anx7688",
                 .of_match_table = anx7688_of_match_table,
+		.pm = &anx7688_pm_ops,
         },
         .probe = anx7688_i2c_probe,
         .remove = anx7688_i2c_remove,
-- 
2.34.0

